src/metrics/include/metrics/operation.hpp

```diff
@@ -0,0 +1,260 @@
+    // loop.
+    count_type iters = 0;  // corresponds to the 'n' field in Cedar
+
+    // The number of documents inserted, modified, deleted, etc.
+    count_type ops = 0;  // corresponds to the 'ops' field in Cedar
+
+    // The size in bytes of the number of documents inserted, etc.
+    count_type size = 0;  // corresponds to the 'size' field in Cedar
+
+    // The number of write errors, transient transaction errors, etc. that occurred when performing
+    // the operation. The operation can still be considered OutcomeType::kSuccess even if errors are
+    // reported.
+    count_type errors = 0;  // corresponds to the 'errors' field in Cedar
+
+    // The amount of time it took to perform the operation.
+    Period<ClockSource> duration = {};  // corresponds to the 'duration' field in Cedar
+
+    // Whether the operation succeeded or not.
+    OutcomeType outcome = OutcomeType::kUnknown;  // corresponds to the 'outcome' field in Cedar
+};
+
+
+template <typename ClockSource>
+class OperationImpl {
+public:
+    using time_point = typename ClockSource::time_point;
+    using EventSeries = TimeSeries<ClockSource, OperationEvent<ClockSource>>;
+
+    OperationImpl(std::string actorName, std::string opName, EventSeries& events)
+        : _actorName(std::move(actorName)),
+          _opName(std::move(opName)),
+          _events(std::addressof(events)) {}
+
+    /**
+     * @return the name of the actor running the operation.
+     */
+    const std::string& getActorName() const {
+        return _actorName;
+    }
+
+    /**
+     * @return the name of the operation being run.
+     */
+    const std::string& getOpName() const {
+        return _opName;
+    }
+
+    void reportAt(time_point finished, OperationEvent<ClockSource>&& event) {
+        _events->addAt(finished, event);
+    }
+
+private:
+    const std::string _actorName;
+    const std::string _opName;
+    EventSeries* const _events;
+};
+
+/**
+ * RAII type for managing data captured about a running operation. Constructing an instance starts a
+ * timer for how long the operation takes. The timer ends when success() or failure() is called on
+ * the instance.
+ */
+template <typename ClockSource>
+class OperationContextT : private boost::noncopyable {
+private:
+    using OutcomeType = typename OperationEvent<ClockSource>::OutcomeType;
+
+public:
+    using time_point = typename ClockSource::time_point;
+
+    explicit OperationContextT(v1::OperationImpl<ClockSource>& op)
+        : _op{std::addressof(op)}, _started{ClockSource::now()} {}
+
+    OperationContextT(OperationContextT<ClockSource>&& other) noexcept
+        : _op{std::move(other._op)},
+          _started{std::move(other._started)},
+          _event{std::move(other._event)},
+          _isClosed{std::exchange(other._isClosed, true)} {}
+
+    ~OperationContextT() {
+        if (!_isClosed) {
+            BOOST_LOG_TRIVIAL(warning)
```

Comment:
maybe `error`? it is programmer-error that affects data-quality.

Reply:
Done.